src/dlfcn.c

```diff
@@ -467,7 +472,174 @@ char *dlerror( void )
     return error_buffer;
 }
 
-#ifdef SHARED
+/* taken from http://bandido.ch/programming/Import_Address_Table_Hooking.pdf */
+static IMAGE_IMPORT_DESCRIPTOR* get_import_address_table( HMODULE module, DWORD *size )
+{
+   IMAGE_DOS_HEADER *dosHeader;
+   IMAGE_OPTIONAL_HEADER *optionalHeader;
+
+   dosHeader = (IMAGE_DOS_HEADER *) module;
+   if( dosHeader->e_magic != 0x5A4D )
+       return NULL;
+   optionalHeader = (IMAGE_OPTIONAL_HEADER *) ((BYTE *) module + dosHeader->e_lfanew + 24);
+    if ( optionalHeader->Magic != IMAGE_NT_OPTIONAL_HDR_MAGIC )
+        return NULL;
+    if ( optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].Size == 0 ||
+        optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].VirtualAddress == 0 )
+        return NULL;
+    *size = optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].Size;
+    return (IMAGE_IMPORT_DESCRIPTOR*)((BYTE*)module +
+        optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].VirtualAddress);
+}
+
+/*
+ * return symbol name for a given address
+ */
+static const char *get_symbol_name( HMODULE module, IMAGE_IMPORT_DESCRIPTOR *iid, void *addr )
+{
+    int i;
+    for(i = 0; iid[i].Characteristics != 0 && iid[i].FirstThunk != 0; i++) {
+        BYTE *base = (BYTE *)module; /* required to have correct calculations */
+        IMAGE_THUNK_DATA *thunkILT = (IMAGE_THUNK_DATA*)(iid[i].Characteristics + base);
+        IMAGE_THUNK_DATA *thunkIAT = (IMAGE_THUNK_DATA*)(iid[i].FirstThunk + base);
+        for(; thunkILT->u1.AddressOfData != 0; thunkILT++, thunkIAT++) {
+            IMAGE_IMPORT_BY_NAME *nameData;
+            if (IMAGE_SNAP_BY_ORDINAL(thunkILT->u1.Ordinal))
+              continue;
+            if ((void *)thunkIAT->u1.Function != addr)
+              continue;
+            nameData = (IMAGE_IMPORT_BY_NAME*)(thunkILT->u1.AddressOfData + base);
+            return (const char *)nameData->Name;
+        }
+   }
+   return NULL;
+}
+
+static BOOL is_valid_address( void *addr )
+{
+    if( addr == NULL )
+        return FALSE;
+
+    /* check valid pointer */
+    MEMORY_BASIC_INFORMATION info;
+    SIZE_T result;

```

Comment:
Same problem as here: https://github.com/dlfcn-win32/dlfcn-win32/pull/72#pullrequestreview-550934633

Reply:
Still not resolved, variables needs to be delared at beginning, prior code `if( addr == NULL )`.

Reply:
Due to this issue compilation under WDK is failing on error:
```
dlfcn.c(545) : error C2275: 'MEMORY_BASIC_INFORMATION' : illegal use of this type as an expression
```